REGRESSION (274086@main?): ASSERTION FAILED: lock.isHeld() in animationsWereRemovedFromNode() on http/tests/site-isolation/iframe-process-termination-after-navigation-completed.html

graouts opened the pull request:
#### 74c0aa4209e14edf714b16d023ac838b718120f8
<pre>
REGRESSION (274086@main?): ASSERTION FAILED: lock.isHeld() in animationsWereRemovedFromNode() on http/tests/site-isolation/iframe-process-termination-after-navigation-completed.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=269251">https://bugs.webkit.org/show_bug.cgi?id=269251</a>
<a href="https://rdar.apple.com/122838302">rdar://122838302</a>

Reviewed by Cameron McCormack.

For atomicity, the transaction handling code was made to take the lock for the whole transaction
in `RemoteScrollingCoordinatorProxyMac::willCommitLayerAndScrollingTrees()` in 273391@main. Since
`RemoteLayerTreeHost::remotePageProcessCrashed()` is a pseudo-transaction, it should also take
the lock.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::remotePageProcessCrashed):

Canonical link: <a href="https://commits.webkit.org/274882@main">https://commits.webkit.org/274882@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9f9ea6d9882bd93d121c36648e05898d07fa49f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40159 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19171 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42537 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42704 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22094 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16500 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33396 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40733 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16154 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34684 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13958 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14002 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35636 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43982 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36435 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35970 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39720 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14974 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12281 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37997 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16593 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16642 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5328 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16237 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -107,8 +107,14 @@
+    if (!m_remoteLayerTreeHost)
+        return;
+
+    if (auto* scrollingCoordinator = m_webPageProxy->scrollingCoordinatorProxy()) {
```
Very small nit: other code in this file does `auto scrollingCoordinator = ...`, i.e. without the `*`.

cdumez replied:
We usually prefer `auto*` for raw pointers though.

cdumez replied:
> We usually prefer `auto*` for raw pointers though.

I think it's OK for new code to not match existing one if this is a positive change.

heycam commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -107,8 +107,14 @@
+        return;
+
+    if (auto* scrollingCoordinator = m_webPageProxy->scrollingCoordinatorProxy()) {
+        scrollingCoordinator->willCommitLayerAndScrollingTrees();
         m_remoteLayerTreeHost->remotePageProcessCrashed(processIdentifier);
+        scrollingCoordinator->didCommitLayerAndScrollingTrees();
```
Can we ever be in a situation where `scrollingCoordinator` is null, but `m_remoteLayerTreeHost` is non-null, and we still need to call `remotePageProcessCrashed`?

smfr replied:
This code reads oddly; the willCommitLayerAndScrollingTrees/didCommitLayerAndScrollingTrees pair aren't surrounding a real commit here, so we should rename them, or otherwise refactor.

graouts replied:
I think we can land as-is to restore the test and fix the crash on the bots and follow up with an improved naming or pattern as soon as possible.

graouts committed:
REGRESSION (274086@main?): ASSERTION FAILED: lock.isHeld() in animationsWereRemovedFromNode() on http/tests/site-isolation/iframe-process-termination-after-navigation-completed.html
https://bugs.webkit.org/show_bug.cgi?id=269251
rdar://122838302

Reviewed by Cameron McCormack.

For atomicity, the transaction handling code was made to take the lock for the whole transaction
in `RemoteScrollingCoordinatorProxyMac::willCommitLayerAndScrollingTrees()` in 273391@main. Since
`RemoteLayerTreeHost::remotePageProcessCrashed()` is a pseudo-transaction, it should also take
the lock.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::remotePageProcessCrashed):

Canonical link: https://commits.webkit.org/274882@main
